### PR TITLE
Add helpful warnings for empty Epochs

### LINKED ISSUE #11305: Non-helping Error message for methods on empty epochs
### Description of the problem

This is an issue I saw often with beginner users of mne-python:
When an `Epochs`-instance is created and for some reason all epochs are marked as bad (reject/flat-parameter, annotations), the following methods (among others) throw error messages, which are especially for beginners not really helpful:

- `average`
- `plot`
- `plot_image`
- `plot_psd`
- `plot_topo_image`

When epochs are manually dropped with `Epochs.drop_bad()` or when they are dropped when data is loaded, there is no warning message indicating that all available epochs have been dropped. The only part I found with a warning is when `Epochs.save()` is used on an empty Epochs-instance.

### Solution
I would suggest to warn when all epochs are about to be dropped and to capture the exceptions due to empty Epochs-instances inside the relevant methods with more expressive messages.

### Steps to reproduce

```Python
import os

import mne

sample_data_folder = mne.datasets.sample.data_path()
sample_data_raw_file = (sample_data_folder / 'MEG' / 'sample' /
                        'sample_audvis_raw.fif')
raw = mne.io.read_raw_fif(sample_data_raw_file, verbose=False)

sample_data_events_file = os.path.join(sample_data_folder,
                                       'MEG', 'sample',
                                       'sample_audvis_raw-eve.fif')
events = mne.read_events(sample_data_events_file)
epochs = mne.Epochs(raw, events, tmin=-0.3, tmax=0.7, reject={'eog': 1e-10},
                    reject_by_annotation=False)

# All epochs dropped, no warning message
epochs.drop_bad()

# ZeroDivisionError: float division by zero
epochs.average()

# ZeroDivisionError: float division by zero
epochs.plot()

# ValueError: zero-size array to reduction operation maximum which has no identity
epochs.plot_image()

# ValueError: zero-size array to reduction operation minimum which has no identity
epochs.plot_psd()

# ValueError: zero-size array to reduction operation maximum which has no identity
epochs.plot_topo_image()
```

### Link to data

_No response_

### Expected results

More concisive warnings and error messages helping especially beginners with finding the problem.

### Actual results

`ZeroDivisionError: float division by zero`

`ValueError: zero-size array to reduction operation maximum which has no identity`

### Additional information

Platform:         Windows-10-10.0.22000-SP0
Python:           3.10.5 | packaged by conda-forge | (main, Jun 14 2022, 06:57:19) [MSC v.1929 64 bit (AMD64)]
Executable:       \anaconda3\envs\mnedev\python.exe
CPU:              Intel64 Family 6 Model 158 Stepping 9, GenuineIntel: 8 cores
Memory:           15.9 GB
mne:              1.3.dev0
numpy:            1.22.4 {unknown linalg bindings}
scipy:            1.9.0
matplotlib:       3.5.2Backend QtAgg is interactive backend. Turning interactive mode on.
 {backend=QtAgg}
sklearn:          1.1.2
numba:            0.55.2
nibabel:          4.0.1
nilearn:          0.9.1
dipy:             1.5.0
openmeeg:         Not found
cupy:             Not found
pandas:           1.4.3
pyvista:          0.36.1 {OpenGL 4.4.0 - Build 21.20.16.4526 via Intel(R) HD Graphics 630}
pyvistaqt:        0.9.0
ipyvtklink:       0.2.2
vtk:              9.1.0
qtpy:             2.1.0 {PyQt5=5.15.4}
ipympl:           0.9.1
pyqtgraph:        0.12.4
pooch:            v1.6.0
mne_bids:         0.11.dev0
mne_nirs:         Not found
mne_features:     Not found
mne_qt_browser:   0.4.dev0
mne_connectivity: Not found
mne_icalabel:     Not found

### PR DESCRIPTION
**Pull request by marsipu**
#### Reference issue
Fixes #11305

#### What does this implement/fix?
This PR adds a warning when all epochs are dropped from an `Epochs`-instance.
Furthermore it raises a `RuntimeError` when certain methods are called on empty `Epochs`.

#### Code for testing
```python
import inspect
import os
import traceback

import matplotlib
import mne
import numpy as np
from matplotlib.figure import Figure

matplotlib.use("agg")

mne.set_log_level("WARNING")

sample_data_folder = mne.datasets.sample.data_path()
sample_data_raw_file = (sample_data_folder / "MEG" / "sample" /
                        "sample_audvis_raw.fif")
raw = mne.io.read_raw_fif(sample_data_raw_file, verbose=False)
sample_data_events_file = os.path.join(sample_data_folder,
                                       "MEG", "sample",
                                       "sample_audvis_raw-eve.fif")
events = mne.read_events(sample_data_events_file)
tmin = -0.3
tmax = 0.7
epochs = mne.Epochs(raw, events, tmin=tmin, tmax=tmax, reject_by_annotation=False)
epochs_empty = mne.Epochs(raw, events, tmin=tmin, tmax=tmax, reject={"eog": 1e-10}, reject_by_annotation=False)

# Create preloaded epochs
epochs_preloaded = epochs.copy().load_data()
epochs_empty_preloaded = epochs_empty.copy().load_data()

# Drop bad epochs
epochs.drop_bad()
epochs_empty.drop_bad()
epochs_preloaded.drop_bad()
epochs_empty_preloaded.drop_bad()

def test_fun(data):
    return data * 2


arg_dict = {
    "add_channels": {"add_list": ["Test1"]},
    "add_proj": {"projs": mne.read_proj(os.path.join(sample_data_folder, "MEG", "sample", "sample_audvis_ecg-proj.fif"))},
    "add_reference_channels": {"ref_channels": "EEG 999"},
    "apply_function": {"fun": test_fun},
    "decimate": {"decim": 2},
    "del_proj": {"idx": 0},
    "drop": {"indices": [0]},
    "drop_channels": {"ch_names": ["MEG 0111"]},
    "export": {"fname": "test.set", "overwrite": True},
    "filter": {"l_freq": 1, "h_freq": 40},
    "pick": {"picks": ["eeg"]},
    "pick_channels": {"ch_names": ["MEG 0111"]},
    "pick_types": {"meg": True},
    "plot_psd_topo": {"tmin": 0.1, "tmax": 0.2},
    "rename_channels": {"mapping": {"MEG 0111": "Test1"}},
    "reorder_channels": {"ch_names": epochs.ch_names[::-1]},
    "resample": {"sfreq": 100},
    "save": {"overwrite": True},
    "savgol_filter": {"h_freq": 40},
    "set_annotations": {"annotations": mne.Annotations(onset=[0.1], duration=[0.2], description=["test"])},
    "set_channel_types": {"mapping": {"MEG 0111": "eeg"}},
    "set_meas_date": {"meadate": epochs.info["meas_date"]},
    "shift_time": {"tshift": 0.1},
    "subtract_evoked": {"evoked": epochs.average()},
    "time_as_index": {"times": epochs.times[0]},
}

# Check every method
for meth_name in dir(epochs):
    if not meth_name.startswith("_"):
        meth = getattr(epochs, meth_name)
        if inspect.ismethod(meth):
            print(f"\n############# Executing {meth_name}...##############")
            for epo_name, epo in [("Epochs (unpreloaded)", epochs),
                                  ("Epochs (loaded)", epochs_preloaded),
                                  ("Empty epochs (unpreloaded)", epochs_empty),
                                  ("Empty epochs (loaded)", epochs_empty_preloaded)]:

                print(f"---------------------------------------------\n{epo_name}")
                try:
                    kwargs = arg_dict.get(meth_name, dict())
                    if meth_name == "save":
                        kwargs["fname"] = f"{epo_name}_test-epo.fif"
                    result = getattr(epo.copy(), meth_name)(**kwargs)
                except Exception as e:
                    if "By default, MNE does not load data" in str(e):
                        print("Doesn't run unpreloaded")
                    elif "can't run because this Epochs-object is empty." in str(e):
                        print("Handling the exception worked!!")
                    else:
                        print(traceback.format_exc())
                else:
                    # Close plots
                    getattr(result, "close", lambda: None)()
                    if isinstance(result, Figure):
                        matplotlib.pyplot.close(result)
```

### CODE REVIEW

**Comment by marsipu** [comment]
@agramfort With these changes the relevant methods are prevented from running when `Epochs` is empty.
`Epochs.save()` only prints a warning and allows saving empty `Epochs`. Is this intentional or should `Epochs.save()` also not run if `Epochs` is empty?

File: mne/epochs.py [hunk 1]
```diff
@@ -338,6 +339,26 @@ def _handle_event_repeated(events, event_id, event_repeated, selection,
...
+    def wrapper(*args, **kwargs):
+        # Prevent method from running if epochs are empty
+        if len(args[0]) == 0:
+            warn(f'{meth.__name__} can not run because Epochs are empty!')
+        else:
+            # Catch exceptions when epochs are dropped
+            # during execution of the method
+            try:
+                return meth(*args, **kwargs)
+            except (ZeroDivisionError, ValueError, IndexError):
```
Comment (agramfort) [inline mne/epochs.py:353]: I dont like much exceptions that hide where exactly the problem is. I would rather be explicit in the wrapper. If it crashes for empty epochs rather than catching an exception we should just immediately raise one. It makes no sense to compute the psd of empty epochs for example.


Also can I ask you to add a tests calling all wrapped methods on some empty epochs?

🙏

Comment (larsoner) [reply mne/epochs.py:353]: > It makes no sense to compute the psd of empty epochs for example.

The tricky part here is that things like `len(epochs)` / epoch count only exist once bads have been dropped, which sometimes isn't available until *after* running the method that iterates over the epochs to see if they're valid (i.e., with non-preloaded epochs).

Naively we could do `self.drop_bad()` and then check, but this is inefficient because it could cause iteration over epochs (reading from disk!) twice instead of once.

One safe way out of this I think might be to:

1. change the `__iter__` method to have a `raise_empty=False`, that these methods should set to True if no valid epochs are found
2. change the `_check_preload` function to have a `raise_empty=False` as well
3. ensure that any methods that need non-empty Epochs either iterate using `__iter__(..., raise_empty=True)` or call `_check_preload(..., raise_empty=True)

Something like this I think could work

Comment (marsipu) [reply mne/epochs.py:353]: > Also can I ask you to add a tests calling all wrapped methods on some empty epochs?

Yes

>The tricky part here is that things like len(epochs) / epoch count only exist once bads have been dropped, which sometimes isn't available until after running the method that iterates over the epochs to see if they're valid (i.e., with non-preloaded epochs).

Yes, I think this is also why the tests are failing at the moment.

> 1. change the __iter__ method to have a raise_empty=False, that these methods should set to True if no valid epochs are found

@larsoner Just curious, would there be a way to pass `raise_empty=True` into an iteration call like `for epo in Epochs()` too/use `iter()`? Or does it have to be `__iter__` then?

Comment (larsoner) [reply mne/epochs.py:353]: I think you have to use `__iter__` directly, which should be fine

Comment (marsipu) [reply mne/epochs.py:392]: > 2. change the _check_preload function to have a raise_empty=False as well

I am not sure if this is necessary, since only `BaseEpochs.crop` and `BaseEpochs.apply_function` are using `_check_preload` and this helper-function is used by a wide range of object-types to have this specific signature change. WDYT?

**Comment by larsoner** [comment]
@marsipu I got an email this morning that you commented on this but now I don't see your comment. Did you delete it or something? If not then I can (re-)post it here and reply, or feel free to reformulate your thought and post again but that is more work!

**Comment by marsipu** [comment]
@larsoner Yes, I was getting into this issue again and I got stuck and posed a question which I later deleted because it was outdated. Sorry, didn't think about that you were notified immediately.
I am still stuck.
Did I understand correctly that you suggest replacing the for-loops in each method with `__iter__(check_empty=True)` if they should also work with unpreloaded epochs?
I have troubles finding where if and where a for-loop exists (e.g. in BaseEpochs.drop -> GetEpochs.Mixin._getitem()). And seems to me a lot of very specific changes for this general problem.

File: mne/epochs.py (hunk 1 above)
Comment (marsipu) [reply mne/epochs.py:392]: Maybe I understand something wrong here, but there seems to be a paradox for me:
__iter__ is at the beginning when we iterate the Epochs-object to check if it has any epochs, right?
How can we raise empty in this method then if we haven't iterated over it yet?

**Comment by marsipu** [comment]
Another solution (for unpreloaded epochs) could maybe be:
When the while loop in GetEpchsMixin.__next__ rushes through without any good epochs, we could raise an Error there if an internal attribute (e.g. BaseEpochs._raise_empty) is True which can be set by the decorator or in the concerning method itself.
And for preloaded we can simply check _data length.
@larsoner WDYT?

File: mne/epochs.py (hunk 1 above)
Comment (marsipu) [reply mne/epochs.py:392]: @larsoner In which case do we have empty epochs without having called `Epochs.drop_bad()` internally or externally? (isn't this the most common reason for empty eochs?)

**Comment by marsipu** [comment]
A general question: The PR/commits seem to lack a status to run all the ubuntu-CIs. Is it something you have to enable or do I have to follow a naming convention in the commits?

File: mne/epochs.py (hunk 1 above)
Comment (larsoner) [reply mne/epochs.py]: @marsipu I think you miss the point a bit here about `epochs.drop_bad` / `epochs.drop`. I am talking about how any "empty" check needs to be implemented. There are two ways I know of that our methods that operate on epochs currently get the data they need:

1. They iterate over epochs and take the data from each and use it in some way via `for epoch in epochs` or so. Hence my suggestion that, for such functions, being able to do `for epoch in epochs.__iter__(raise_error=True)` might be useful for some of them and `raise_error=False` might be better in others, depending on whether or not a sensible value can be returned in the case of empty epochs.
2. `_check_preload` then use `_data` directly -- hence my point (2) above about modifying `_check_preload` to look for emptiness for functions where empty epochs would again lead to something that doesn't make sense..

And then point (3) is just a suggestion to use these two where needed.

**Comment by larsoner** [comment]
> Is it something you have to enable as maintainers or do I have to follow a naming convention in the commits?

GitHub is having a bad day this morning so I'll blame that. It looks like you did everything fine 🤷

**Comment by larsoner** [comment]
> When the while loop in GetEpchsMixin.__next__ rushes through without any good epochs, we could raise an Error there if an internal attribute (e.g. BaseEpochs._raise_empty) is True which can be set by the decorator or in the concerning method itself.

I would rather just explicitly use `raise_empty` inside the given method. Using a decorator to do this is more implicit (and hence harder to understand) even though it would be a neat trick :)

**Comment by marsipu** [comment]
@larsoner I did a thorough analysis of every method of `Epochs` now (see Code in first comment).
With `_check_preload` some errors can be captured with `__iter__` only a few and only when epochs are preloaded.
The following problematic methods remain:

<details>
<summary>Expand</summary>
## apply_hilbert
_check_preload-Fix: :white_check_mark:

## as_type
ValueError: zero-size array to reduction operation maximum which has no identity

cause: events with length=0

_check_preload-Fix: :x:

__iter__-Fix: :x:

## average
ZeroDivisionError: float division by zero

Cause: events with length=0

_check_preload-Fix: :x:

__iter__-Fix: (:white_check_mark:) only for unpreloaded

## compute_psd, plot_psd, plot_psd_topo, plot_psd_topomap
ValueError: zero-size array to reduction operation minimum which has no identity

Cause: _get_data() returns array with size 0 in epoch-dimension

_check_preload-Fix: :x:

__iter__-Fix: :x:

## plot
IndexError: index 0 is out of bounds for axis 0 with size 0

Cause: boundary_times created on events with length=0

_check_preload-Fix: :x:

__iter__-Fix: :x:

## plot_image
ValueError: zero-size array to reduction operation maximum which has no identity

Cause: events with length=0

_check_preload-Fix: :x:

__iter__-Fix: :x:

## plot_psd
ValueError: zero-size array to reduction operation minimum which has no identity

Cause: _data-array with size 0 in epoch-dimension

_check_preload-Fix: :x:

__iter__-Fix: :x:

## plot_topo_image
ValueError: x and y must have same first dimension, but have shapes (101,) and (364, 101)

Cause: _data-array with size 0 in epoch-dimension (which later causes empty _data-array in psd-dimension)

_check_preload-Fix: :x:

__iter__-Fix: :x:

## standard_error
ZeroDivisionError: float division by zero

Cause: events with length=0

_check_preload-Fix: :x:

__iter__-Fix: (:white_check_mark:) only for unpreloaded

## to_data_frame
ValueError: need at least one array to concatenate

Cause: _get_data() returns array with size 0 in epoch-dimension

_check_preload-Fix: :x:

__iter__-Fix: :x:
</details>

How do you think we should deal with these? Having a decorator or something at method-level seems more readable to me than checking in each (often also shared) sub-method.

File: mne/utils/check.py [hunk 2]
```diff
@@ -301,13 +301,15 @@ def _check_preload(inst, msg):
...
         name = "epochs" if isinstance(inst, BaseEpochs) else "raw"
-        if not inst.preload:
+        if not getattr(inst, "preload", False):
             raise RuntimeError(
                 "By default, MNE does not load data into main memory to "
                 "conserve resources. " + msg + " requires %s data to be "
                 "loaded. Use preload=True (or string) in the constructor or "
                 "%s.load_data()." % (name, name)
             )
+        elif name == "epochs" and len(inst) == 0:
```
Comment (marsipu) [inline mne/utils/check.py:311]: @larsoner The usage from epochs seems to be little compared to the other instances that share this method. And I don't see a case where we would check preload and accept empty epochs. Do you think an extra-arg is necessary?

Comment (larsoner) [reply mne/utils/check.py:311]: You could try without adding the new kwarg and see if it works

**Comment by larsoner** [comment]
> How do you think we should deal with these? Having a decorator or something at method-level seems more readable to me than checking in each (often also shared) sub-method.

Rather than a decorator I would add a `_check_empty(self)` or something like we have for `_check_preload`. It's less magical than a decorator.

To me all of those methods could probably just throw an error on empty epochs. The only exception is maybe `to_data_frame`, but we can add the exception-on-empty now and relax it later if someone really needs it

**Comment by marsipu** [comment]
@larsoner I saw that you have an extra test to allow empty epochs to be saved.
I am wondering why, since `mne.read_epochs` throws an error when you try to read an empty epoch again. I left `save` out of the empty-checks for now.

File: mne/epochs.py [hunk 3]
```diff
@@ -1166,6 +1169,7 @@ def _compute_aggregate(self, picks, mode="mean"):
                 )
             data = np.zeros((n_channels, n_times))
             n_events = 0
+            # __iter__ here??
             for e in self:
```
Comment (larsoner) [inline mne/epochs.py:1173]: Yes, this is the sort of place where I think you'd want
```suggestion
            # __iter__ here??
            for e in self.__iter__(raise_error=True):
```

File: mne/epochs.py [hunk 4]
```diff
@@ -1741,7 +1748,17 @@ def _get_data(
             del drop_log
 
             self._bad_dropped = True
-            logger.info("%d bad epochs dropped" % (n_events - len(good_idx)))
+            n_bads_dropped = n_events - len(good_idx)
+            logger.info(f"{n_bads_dropped} bad epochs dropped")
+
+            if n_bads_dropped == n_events:
+                logger.warning(
```
Comment (larsoner) [inline mne/epochs.py:1755]: Don't use `logger.warning` but instead `mne.utils.warn(...)`, it does some nice `verbose`-based magic

File: mne/epochs.py [hunk 5]
```diff
@@ -2600,6 +2621,7 @@ def as_type(self, ch_type="grad", mode="fast"):
         """
         from .forward import _as_meg_type_inst
 
+        _check_empty(self, "as_type")
         return _as_meg_type_inst(self, ch_type=ch_type, mode=mode)
```
Comment (larsoner) [inline mne/epochs.py:2625]: Same, this could call `get_data(..., on_empty='raise')`.

It makes me think that we might want to add `get_data(..., on_missing='raise')` by default as a safety/bugfix, then we get all of these checks for free in our code base and users probably get safer behavior (arguably cannot `get_data` if there is none to get...) with the ability to do `on_empty='ignore'` if they *really* want an array of size 0

File: mne/filter.py [hunk 6]
```diff
@@ -2498,6 +2499,7 @@ def savgol_filter(self, h_freq, verbose=None):
         from scipy.signal import savgol_filter
 
         _check_preload(self, "inst.savgol_filter")
+        _check_empty(self, "savgol_filter")
```
Comment (larsoner) [inline mne/filter.py:2502]: ... and then the common `_check_preload() _check_empty()` pattern makes me think we should have `_check_preload` *always* just check for emptiness, too. That will probably satisfy 90-100% of our internal uses of `_check_preload` for epochs and if there are 10% where we *do* want to allow preload-and-no-epochs then we can add `_check_preload(..., allow_empty=False)` kwarg and pass `allow_empty=True` to the ones that need it.

**Review by larsoner — COMMENTED** [review]
The diff isn't bad! But I'm thinking we might be able to streamline it further, see inline comments. The TL;DR is I suggest:

1. Modify `epochs.get_data(..., on_empty='raise')` by default as a bugfix. Internally we can use `on_empty='ignore'` if there are places we really want to ignore it.
2. Modify `_check_preload` to do the emptiness check itself. Hopefully no new kwarg is needed but if it is we can add `_check_preload(..., allow_empty=False)` that we can pass `allow_empty=True` in the few places we need it.

Regarding the ability to save but not load empty epochs, this does not really make any sense. I think I'd prefer to allow loading empty ones if it's not too much work. If it is a lot of work you can just leave it with a code comment marked `TODO`. Maybe we'll get rid of `save` for empty epochs someday but that would require more thought/decisions probably left to another PR.

**Comment by larsoner** [comment]
... and @marsipu let me know if you get tired of making these changes and I can give it a shot. But happy to wait for you to tackle it if you want, no rush from our end :+1:

**Comment by marsipu** [comment]
@larsoner thanks for the feedback.
I am on vacation for one week now, but I'd be back afterwards to finalize this PR with your suggestions. It would be satisfying to finally solve this sufficiently:grin:

File: mne/epochs.py (hunk 4 above)
Comment (marsipu) [reply mne/epochs.py:1755]: I am probably making a mistake here, but I can't get my new test to pass. They always fail on the RuntimeWarning and with `pytest.warns(RuntimeWarning)` it fails anyway at `warn_explicit`. @larsoner Can you help me there?

File: mne/epochs.py [hunk 7]
```diff
@@ -2529,7 +2562,7 @@ def to_data_frame(
         time_format = _check_time_format(time_format, valid_time_formats)
         # get data
         picks = _picks_to_idx(self.info, picks, "all", exclude=())
-        data = self.get_data()[:, picks, :]
+        data = self._get_data(on_empty="raise")[:, picks, :]
```
Comment (marsipu) [inline mne/epochs.py:2565]: I was reluctant to change the signature of public methods, so I sometimes switched to the private method. WDYT?

File: mne/tests/test_epochs.py [hunk 8]
```diff
@@ -4795,3 +4795,45 @@ def test_epochs_saving_with_annotations(tmp_path):
...
+    arg_values = [(k, v) for k, v in test_methods.items()]
+    arg_ids = test_methods.keys()
+    return {"argnames": "method", "argvalues": arg_values, "ids": arg_ids}
+
+
+@pytest.mark.parametrize(**_get_empty_parametrize())
+def test_empty_error(method, epochs_empty):
+    """Test that a helpful error is raised when certain methods are called."""
+    # Test that a RuntimeError is raised when certain methods are called
+    with pytest.warns(RuntimeWarning, match="dropped"):
```
Comment (marsipu) [inline mne/tests/test_epochs.py:4837]: This is probably where I get something wrong. It just doesn't seem to be catching the `RuntimeWarning `from `mne.utils.warn`

**Comment by marsipu** [comment]
@larsoner First of all thanks a lot for your great idea with the `_get_data`, made a lot of sense in the end.:blush:
I added some comments for you.

File: mne/epochs.py (hunk 4 above)
Comment (marsipu) [reply mne/epochs.py:1756]: Other tests also seem to be affected if I replace `logger.warning` with `warn`

File: mne/tests/test_epochs.py [hunk 9]
```diff
@@ -2291,7 +2291,7 @@ def test_crop(tmp_path):
         1,
         picks=picks,
         preload=True,
-        reject=reject,
+        reject=None,
```
Comment (marsipu) [inline mne/tests/test_epochs.py:2294]: I added this because otherwise it would produce an empty epoch

File: mne/tests/test_epochs.py [hunk 10]
```diff
@@ -4280,7 +4280,7 @@ def test_pick_types_reject_flat_keys():
         event_id,
         preload=True,
         picks=picks,
-        reject=dict(grad=1e-10, mag=1e-10, eeg=1e-3, eog=1e-3),
+        reject=dict(grad=1e-9, mag=1e-10, eeg=1e-3, eog=1e-3),
```
Comment (marsipu) [inline mne/tests/test_epochs.py:4283]: Same here

**Comment by marsipu** [comment]
Using `@pytest.mark.filterwarnings` seems to work. Is this the way?

File: mne/epochs.py [hunk 11]
```diff
@@ -467,12 +467,18 @@ def __init__(
...
         if events is not None:  # RtEpochs can have events=None
             events = _ensure_events(events)
-            events_max = events.max()
-            if events_max > INT32_MAX:
-                raise ValueError(
-                    f"events array values must not exceed {INT32_MAX}, "
-                    f"got {events_max}"
-                )
+            # Allow reading empty epochs (ToDo: Maybe not anymore in the future)
+            if events.shape == (0, 3):
```
Comment (larsoner) [inline mne/epochs.py:471]: This is more explicit about what you're checking. You could also use `if events.size == 0` or `events.shape[0] == 0` if you want. (It makes it so you don't have to think "oh this could also evaluate to False if events.shape[1] is not equal to 3" and then realize that could never happen based on prior code. So cleaner just to check the thing you're actually interested in and ignore `shape[1]` entirely.)
```suggestion
            if len(events) == 0:
```

File: mne/epochs.py [hunk 12]
```diff
@@ -1560,6 +1569,15 @@ def _project_epoch(self, epoch):
             epoch = np.dot(self._projector, epoch)
         return epoch
 
+    def _handle_empty(self, on_empty, msg):
```
Comment (larsoner) [inline mne/epochs.py:1572]:
```suggestion
    def _handle_empty(self, on_empty, meth):
```

File: mne/epochs.py [hunk 13]
```diff
@@ -1560,6 +1569,15 @@ def _project_epoch(self, epoch):
...
+        if len(self.events) == 0:
+            msg = (
+                f"epochs.{msg}() can't run because this Epochs-object is empty.\n"
```
Comment (larsoner) [inline mne/epochs.py:1575]: Remove newline (and use new var) as it seems unnecessary here
```suggestion
                f"epochs.{meth}() can't run because this Epochs-object is empty. "
```

File: mne/epochs.py [hunk 14]
```diff
@@ -1728,7 +1749,17 @@ def _get_data(
...
-            logger.info("%d bad epochs dropped" % (n_events - len(good_idx)))
+            n_bads_dropped = n_events - len(good_idx)
+            logger.info(f"{n_bads_dropped} bad epochs dropped")
+
+            if n_bads_dropped == n_events:
+                warn(
+                    "All epochs were dropped!\n"
+                    "You might need to alter reject/flat-criteria "
+                    "or drop bad channels to avoid this.\n"
+                    "Tip: With Epochs.plot_drop_log() you can see which "
```
Comment (larsoner) [inline mne/epochs.py:1760]:
```suggestion
                    "or drop bad channels to avoid this. "
                    "You can use Epochs.plot_drop_log() to see which "
```

File: mne/epochs.py [hunk 15]
```diff
@@ -3523,8 +3555,15 @@ def _read_one_epoch_file(f, tree, preload):
...
-
+        try:
+            events, mappings = _read_events_fif(fid, tree)
+        except ValueError as e:
+            # Allow reading empty epochs (ToDo: Maybe not anymore in the future)
+            if str(e) == "Could not find any events":
+                events = np.empty((0, 3), dtype=np.int32)
+                mappings = dict()
+            else:
+                raise e
```
Comment (larsoner) [inline mne/epochs.py:3566]: A bare `raise` behaves better than `raise e` (it makes it omit the current line in the traceback)
```suggestion
                raise
```

File: mne/tests/test_epochs.py [hunk 16]
```diff
@@ -346,6 +346,7 @@ def test_hierarchical():
 
 
 @pytest.mark.slowtest
+@pytest.mark.filterwarnings("ignore:All epochs were dropped!.*:RuntimeWarning")
```
Comment (larsoner) [inline mne/tests/test_epochs.py:349]: Rather than filtering for the entire test, it would be better to target the hopefully one or two lines that actually produce no epochs. If it's more than that number of calls the test is probably buggy and I can take a look.

File: mne/utils/check.py [hunk 17]
```diff
@@ -300,13 +300,15 @@ def _check_preload(inst, msg):
         pass
     else:
         name = "epochs" if isinstance(inst, BaseEpochs) else "raw"
-        if not inst.preload:
+        if not getattr(inst, "preload", False):
```
Comment (larsoner) [inline mne/utils/check.py:303]: Why did you need to add this one?

**Review by larsoner — COMMENTED** [review]
Diff looks nice! Mostly minor stuff. Regarding `filterwarnings` see inline comments, TL;DR is that I think we can/should make the catching more targeted with `with pytest.warns(...)` on the (hopefully few) lines that need it.

File: mne/utils/check.py (hunk 17 above)
Comment (marsipu) [reply mne/utils/check.py:303]: There was a test with RawArray where this caused an error

**Comment by marsipu** [comment]
pytest.warns() seems to work for all tests execpt `test_empty_error` which I created. I couldn't understand why but now I see since the other tests seem to work as expected, it probably has something to do with nesting these to pytest-context-managers there with
```python
with pytest.warns(RuntimeWarning, match="were dropped"):
    with pytest.raises(RuntimeError, match="is empty."):
        getattr(epochs_empty.copy(), method[0])(**method[1])
```
Do you know a better and working solution there? Maybe otherwise we can then just leave the pytest.mark there since that seems to work.

**Comment by larsoner** [comment]
> pytest.warns() seems to work for all tests execpt test_empty_error which I created.

I think you had the context manager in the wrong location, see https://github.com/mne-tools/mne-python/pull/11306/commits/88a68fa7c1975d835015e7878f32710d3f3603e5 which makes things pass for me

File: mne/utils/check.py (hunk 17 above)
Comment (larsoner) [reply mne/utils/check.py:303]: I'll revert so I can see the failure... it seems odd and might indicate another bug

Comment (larsoner) [reply mne/utils/check.py]: @marsipu no failures in the test reverting this change, so maybe it wasn't necessary after all? Feel free to add a test case that showcases the issue otherwise

Comment (marsipu) [reply mne/utils/check.py]: Oh okay, can't reproduce the problem either now.

**Comment by larsoner** [comment]
Okay in it goes, thanks @marsipu !
